async_hooks: add AsyncResource.bind utility

jasnell opened the pull request:
Creates an internal AsyncResource and binds a function to it,
ensuring that the function is invoked within execution context
in which bind was called.

This is suggested as a solution for #33723 and is a simple utility over what can already be done with `AsyncResource`.

/cc @addaleax @puzpuzpuz @ronag @Qard @devsnek @targos 

##### Checklist
- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

devsnek commented on lib/async_hooks.js:
```diff
@@ -211,6 +211,13 @@ class AsyncResource {
+  static bind(fn, type = fn?.name) {
+    const asyncResource = new AsyncResource(type || 'bound-anonymous-fn');
+    return asyncResource.runInAsyncScope.bind(asyncResource, (...args) => {
+      fn(...args);
```
return value not preserved?

devsnek replied:
actually couldn't this just be `asyncResource.runInAsyncScope.bind(asyncResource, fn)`?

jasnell replied:
Yeah, was premature about opening the PR.. had a few local changes not checked in lol... will get those in soon ;-)

Qard commented:
It would be nice if bind was an instance method so one could easily bind an already created resource and then maybe the static version could just use that internally. 🤔

jasnell commented:
@devsnek ... updated!
@Qard ... good idea, updated!

What becomes obvious is that this is a very simple utility that wraps something that can already be done but eliminates some boilerplate.

devsnek commented on lib/async_hooks.js:
```diff
@@ -211,6 +211,14 @@ class AsyncResource {
+  }
+
+  static bind(fn, type = fn?.name) {
+    return (new AsyncResource(type || 'bound-anonymous-fn')).bind(fn);
```
no points will be deducted for using a temporary variable :P

ronag replied:
parentheses unnecessary?

jasnell replied:
Yeah but stylistically I prefer them (more readable in my opinion)

jasnell commented:
Getting the super duper helpful error...

```
mkdir -p out/doc/api
cp -r doc/api out/doc
if [ -x /home/james/node/node/./node ] && [ -e /home/james/node/node/./node ]; then /home/james/node/node/./node  tools/doc/apilinks.js out/doc/apilinks.json lib/_http_client.js lib/net.js lib/_tls_wrap.js lib/http.js lib/dns.js lib/tls.js lib/path.js lib/domain.js lib/timers.js lib/_http_outgoing.js lib/child_process.js lib/_http_agent.js lib/tty.js lib/querystring.js lib/wasi.js lib/module.js lib/url.js lib/http2.js lib/https.js lib/zlib.js lib/perf_hooks.js lib/_stream_readable.js lib/buffer.js lib/punycode.js lib/console.js lib/trace_events.js lib/cluster.js lib/_stream_duplex.js lib/process.js lib/_http_common.js lib/vm.js lib/stream.js lib/sys.js lib/_http_incoming.js lib/fs.js lib/events.js lib/_stream_passthrough.js lib/_stream_writable.js lib/async_hooks.js lib/readline.js lib/repl.js lib/string_decoder.js lib/constants.js lib/v8.js lib/assert.js lib/_stream_wrap.js lib/util.js lib/_stream_transform.js lib/crypto.js lib/inspector.js lib/_http_server.js lib/os.js lib/dgram.js lib/worker_threads.js lib/_tls_common.js; elif [ -x `which node` ] && [ -e `which node` ] && [ `which node` ]; then `which node`  tools/doc/apilinks.js out/doc/apilinks.json lib/_http_client.js lib/net.js lib/_tls_wrap.js lib/http.js lib/dns.js lib/tls.js lib/path.js lib/domain.js lib/timers.js lib/_http_outgoing.js lib/child_process.js lib/_http_agent.js lib/tty.js lib/querystring.js lib/wasi.js lib/module.js lib/url.js lib/http2.js lib/https.js lib/zlib.js lib/perf_hooks.js lib/_stream_readable.js lib/buffer.js lib/punycode.js lib/console.js lib/trace_events.js lib/cluster.js lib/_stream_duplex.js lib/process.js lib/_http_common.js lib/vm.js lib/stream.js lib/sys.js lib/_http_incoming.js lib/fs.js lib/events.js lib/_stream_passthrough.js lib/_stream_writable.js lib/async_hooks.js lib/readline.js lib/repl.js lib/string_decoder.js lib/constants.js lib/v8.js lib/assert.js lib/_stream_wrap.js lib/util.js lib/_stream_transform.js lib/crypto.js lib/inspector.js lib/_http_server.js lib/os.js lib/dgram.js lib/worker_threads.js lib/_tls_common.js; else echo "No available node, cannot run \"node  tools/doc/apilinks.js out/doc/apilinks.json lib/_http_client.js lib/net.js lib/_tls_wrap.js lib/http.js lib/dns.js lib/tls.js lib/path.js lib/domain.js lib/timers.js lib/_http_outgoing.js lib/child_process.js lib/_http_agent.js lib/tty.js lib/querystring.js lib/wasi.js lib/module.js lib/url.js lib/http2.js lib/https.js lib/zlib.js lib/perf_hooks.js lib/_stream_readable.js lib/buffer.js lib/punycode.js lib/console.js lib/trace_events.js lib/cluster.js lib/_stream_duplex.js lib/process.js lib/_http_common.js lib/vm.js lib/stream.js lib/sys.js lib/_http_incoming.js lib/fs.js lib/events.js lib/_stream_passthrough.js lib/_stream_writable.js lib/async_hooks.js lib/readline.js lib/repl.js lib/string_decoder.js lib/constants.js lib/v8.js lib/assert.js lib/_stream_wrap.js lib/util.js lib/_stream_transform.js lib/crypto.js lib/inspector.js lib/_http_server.js lib/os.js lib/dgram.js lib/worker_threads.js lib/_tls_common.js\""; exit 1; fi;
/home/james/node/node/deps/acorn/acorn/dist/acorn.js:2842
    throw err
    ^

SyntaxError: Unexpected token (219:28)
    at Parser.pp$4.raise (/home/james/node/node/deps/acorn/acorn/dist/acorn.js:2840:15)
    at Parser.pp.unexpected (/home/james/node/node/deps/acorn/acorn/dist/acorn.js:689:10)
    at Parser.pp$3.parseExprAtom (/home/james/node/node/deps/acorn/acorn/dist/acorn.js:2263:12)
    at Parser.pp$3.parseExprSubscripts (/home/james/node/node/deps/acorn/acorn/dist/acorn.js:2092:21)
    at Parser.pp$3.parseMaybeUnary (/home/james/node/node/deps/acorn/acorn/dist/acorn.js:2069:19)
    at Parser.pp$3.parseExprOps (/home/james/node/node/deps/acorn/acorn/dist/acorn.js:2013:21)
    at Parser.pp$3.parseMaybeConditional (/home/james/node/node/deps/acorn/acorn/dist/acorn.js:1996:21)
    at Parser.pp$3.parseMaybeAssign (/home/james/node/node/deps/acorn/acorn/dist/acorn.js:1969:21)
    at Parser.pp$3.parseMaybeConditional (/home/james/node/node/deps/acorn/acorn/dist/acorn.js:2001:30)
    at Parser.pp$3.parseMaybeAssign (/home/james/node/node/deps/acorn/acorn/dist/acorn.js:1969:21) {
  pos: 6131,
  loc: Position { line: 219, column: 28 },
  raisedAt: 6132
}
Makefile:755: recipe for target 'out/doc/apilinks.json' failed
make[1]: *** [out/doc/apilinks.json] Error 1
Makefile:717: recipe for target 'doc-only' failed
make: *** [doc-only] Error 2
```

devsnek commented:
> Getting the super duper helpful error...

We really should fix that, it's stopping us from using all the new syntax :( I had a commit somewhere but I have no idea where it went. You could include a commit that patches this with a try/catch:
https://github.com/nodejs/node/blob/dc00a07426f0b2dee9bfe39c766a703eb9b5b880/tools/doc/apilinks.js#L57-L60

puzpuzpuz commented on doc/api/async_hooks.md:
```diff
@@ -900,12 +920,12 @@ const { createServer } = require('http');
+    // Execution context is bound to the proper outer scope
   }));
+  req.on('close', () => {
+    // Execution context is bound to the scope that cause 'close' to emit.
```
nit: a potential typo: `cause` -> `caused`

jasnell replied:
```suggestion
    // Execution context is bound to the scope that caused 'close' to emit.
```

puzpuzpuz commented on doc/api/async_hooks.md:
```diff
@@ -900,12 +920,12 @@ const { createServer } = require('http');
-    // Prints: true
-    console.log(asyncResource.asyncId() === executionAsyncId());
+  req.on('close', AsyncResource.bind(() => {
+    // Execution context is bound to the proper outer scope
```
nit: missing trailing dot.

puzpuzpuz replied:
nit: not sure if `proper outer` is precise enough here. The `proper` part seems to be misleading. How about just `outer` or `current outer`?

jasnell replied:
```suggestion
    // Execution context is bound to the current outer scope.
```

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/32576/

Flarna commented:
Some comments, nothing blocking and feel free to ignore:

It might be helpful to set `function.length` to the same value as the original function. If used together with `EventEmitter` it's don't care but this API is not limited to this and e.g. a express distinguishes between functions with 3 and 4 arguments.

It may be also helpful to somehow allow to check if a function is bound to an async resource. APMs, OTel plugins,... currently patch `EventEmitter` to wrap listeners for context passing. If this is done already in user code via this API it's not needed anymore.

jasnell commented:
@Flarna ... good suggestions! I've updated to set the `length` property and added an `asyncResource` property to the returned function.

jasnell commented:
Landed in 74df749

Flarna commented on lib/async_hooks.js:
```diff
@@ -211,6 +213,28 @@ class AsyncResource {
+    const ret = this.runInAsyncScope.bind(this, fn);
+    ObjectDefineProperties(ret, {
+      'length': {
+        enumerable: true,
```
I think `length` should use `enumerable: false`

Flarna commented on lib/async_hooks.js:
```diff
@@ -211,6 +213,28 @@ class AsyncResource {
+    if (typeof fn !== 'function')
+      throw new ERR_INVALID_ARG_TYPE('fn', 'Function', fn);
+    const ret = this.runInAsyncScope.bind(this, fn);
+    ObjectDefineProperties(ret, {
```
I think both properties should use `configurable: true`.

Flarna commented:
Seems my comments regarding property descriptor were to late => followup at https://github.com/nodejs/node/pull/34620
